File: rust/src-tauri/src/quota.rs
// Quota fetch + defensive parsing, 1:1 port of QuotaService (SPEC section 16).
// Traps honored here:
//  - server JSON numbers are modeled as strings, numeric fallback tolerated
//  - Extra Usage amountLeft unit is 1e-8 yuan -> cents = (raw + 500000) / 1000000
//  - isEnabled == false must be reported as NotActivated (KimiCodeBar v1.1.1 bug)

use chrono::{DateTime, Local};
use reqwest::Client;
use serde::Serialize;
use serde_json::Value;
use std::sync::OnceLock;
use std::time::Duration;

use crate::credentials;

const USAGES_URL: &str = "https://api.kimi.com/coding/v1/usages";

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct QuotaSegment {
    pub percent: f64,
    pub reset_at: Option<DateTime<Local>>,
}

#[derive(Clone, Copy, Serialize, PartialEq, Eq)]
pub enum ExtraState {
    NotActivated,
    NoData,
    Ready,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExtraInfo {
    pub state: ExtraState,
    pub balance_cents: Option<i64>,
    pub monthly_enabled: bool,
    pub monthly_used_cents: Option<i64>,
    pub monthly_limit_cents: Option<i64>,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct QuotaResult {
    pub five_hour: Option<QuotaSegment>,
    pub week: Option<QuotaSegment>,
    pub extra: Option<ExtraInfo>,
    pub fetched_at: DateTime<Local>,
    pub error: Option<String>,
}

impl QuotaResult {
    fn failed(kind: &str) -> Self {
        QuotaResult {
            five_hour: None,
            week: None,
            extra: None,
            fetched_at: Local::now(),
            error: Some(kind.to_string()),
        }
    }

    /// On failure keep last-known-good data: fill null fields from `last`
    /// (SPEC 16.5 step 2; the UI only shows the failure hint in the status line).
    pub fn fill_missing_from(&mut self, last: &QuotaResult) {
        if self.five_hour.is_none() {
            self.five_hour = last.five_hour.clone();
        }
        if self.week.is_none() {
            self.week = last.week.clone();
        }
        if self.extra.is_none() {
            self.extra = last.extra.clone();
        }
    }
}

fn http_client() -> Option<&'static Client> {
    static CLIENT: OnceLock<Option<Client>> = OnceLock::new();
    CLIENT
        .get_or_init(|| Client::builder().timeout(Duration::from_secs(10)).build().ok())
        .as_ref()
}

pub async fn fetch() -> QuotaResult {
    let Some(token) = credentials::load_token() else {
        return QuotaResult::failed("no-token");
    };
    let Some(client) = http_client() else {
        return QuotaResult::failed("HttpRequestException");
    };
    let resp = match client
        .get(USAGES_URL)
        .header("Authorization", format!("Bearer {token}"))
        .header("Accept", "application/json")
        .send()
        .await
    {
        Ok(r) => r,
        Err(e) => {
            // Keep .NET exception-type names so --test-fetch output matches the WPF reference
            return QuotaResult::failed(if e.is_timeout() {
                "TaskCanceledException"
            } else {
                "HttpRequestException"
            });
        }
    };
    if !resp.status().is_success() {
        return QuotaResult::failed("HttpRequestException");
    }
    let root: Value = match resp.json().await {
        Ok(v) => v,
        Err(_) => return QuotaResult::failed("JsonException"),
    };

    let mut r = QuotaResult {
        five_hour: None,
        week: None,
        extra: None,
        fetched_at: Local::now(),
        error: None,
    };
    // 5-hour segment: root.limits[0].detail
    if let Some(detail) = root
        .get("limits")
        .and_then(|l| l.as_array())
        .and_then(|a| a.first())
        .and_then(|first| first.get("detail"))
    {
        r.five_hour = Some(parse_segment(detail));
    }
    // Week segment: root.usage
    if let Some(u) = root.get("usage") {
        if u.is_object() {
            r.week = Some(parse_segment(u));
        }
    }
    r.extra = Some(parse_extra(root.get("boosterWallet")));
    r
}

/// JSON number-or-string -> f64, missing -> 0.
fn get_f64(v: &Value, key: &str) -> f64 {
    match v.get(key) {
        Some(Value::Number(n)) => n.as_f64().unwrap_or(0.0),
        Some(Value::String(s)) => s.trim().parse::<f64>().unwrap_or(0.0),
        _ => 0.0,
    }
}

/// JSON number-or-string -> i64.
fn get_i64(v: &Value) -> Option<i64> {
    match v {
        Value::Number(n) => n.as_i64(),
        Value::String(s) => s.trim().parse::<i64>().ok(),
        _ => None,
    }
}

/// resetTime: RFC3339 first, then the looser shapes DateTimeOffset.TryParse
/// accepts (space separator, with/without offset; no offset = local time).
fn parse_reset_time(s: &str) -> Option<DateTime<Local>> {
    if let Ok(d) = DateTime::parse_from_rfc3339(s) {
        return Some(d.with_timezone(&Local));
    }
    for fmt in [
        "%Y-%m-%d %H:%M:%S %:z",
        "%Y-%m-%d %H:%M:%S %z",
        "%Y-%m-%dT%H:%M:%S %:z",
        "%Y-%m-%d %H:%M:%S%.f %:z",
    ] {
        if let Ok(d) = DateTime::parse_from_str(s, fmt) {
            return Some(d.with_timezone(&Local));
        }
    }
    for fmt in ["%Y-%m-%d %H:%M:%S", "%Y-%m-%dT%H:%M:%S"] {
        if let Ok(n) = chrono::NaiveDateTime::parse_from_str(s, fmt) {
            if let Some(d) = n.and_local_timezone(Local).single() {
                return Some(d);
            }
        }
    }
    None
}

fn parse_segment(v: &Value) -> QuotaSegment {
    let used = get_f64(v, "used");
    let mut limit = get_f64(v, "limit");
    if limit <= 0.0 {
        limit = 1.0; // guard against division by zero
    }
    let reset_at = v
        .get("resetTime")
        .and_then(|x| x.as_str())
        .and_then(parse_reset_time);
    QuotaSegment {
        percent: used / limit * 100.0,
        reset_at,
    }
}

fn parse_cents(money: Option<&Value>) -> Option<i64> {
    money
        .filter(|m| m.is_object())
        .and_then(|m| m.get("priceInCents"))
        .and_then(get_i64)
}

fn parse_extra(wallet: Option<&Value>) -> ExtraInfo {
    let mut info = ExtraInfo {
        state: ExtraState::NotActivated,
        balance_cents: None,
        monthly_enabled: false,
        monthly_used_cents: None,
        monthly_limit_cents: None,
    };
    let Some(w) = wallet.filter(|w| w.is_object()) else {
        return info; // not an object -> NotActivated
    };

    // isEnabled defense: when the booster is disabled, amountLeft is an estimate
    // (monthly limit minus used), NOT the real balance -> must read as NotActivated.
    if w.get("isEnabled").and_then(|v| v.as_bool()) == Some(false) {
        return info;
    }

    if let Some(raw) = w
        .get("balance")
        .filter(|b| b.is_object())
        .and_then(|b| b.get("amountLeft"))
        .and_then(get_i64)
    {
        info.state = ExtraState::Ready;
        info.balance_cents = Some((raw + 500_000) / 1_000_000); // 1e-8 yuan -> cents, rounded
    } else {
        info.state = ExtraState::NoData;
    }

    if w.get("monthlyChargeLimitEnabled").and_then(|v| v.as_bool()) == Some(true) {
        info.monthly_enabled = true;
        info.monthly_used_cents = parse_cents(w.get("monthlyUsed"));
        info.monthly_limit_cents = parse_cents(w.get("monthlyChargeLimit"));
    }
    info
}

File: qt/src/quota.cpp
#include "quota.h"

#include "credentials.h"
#include "http_util.h"

#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonValue>
#include <QRegularExpression>
#include <QTimeZone>
#include <cmath>
#include <limits>

#ifndef NOMINMAX
#define NOMINMAX
#endif
#include <windows.h>

namespace {

const char kUsagesUrl[] = "https://api.kimi.com/coding/v1/usages";

// Current local time with the sub-second fraction from the precise system
// clock (100ns units), matching chrono::Local::now() on Windows.
void nowPrecise(QDateTime &dt, int &nanos)
{
    FILETIME ft;
    GetSystemTimePreciseAsFileTime(&ft);
    const quint64 ticks = (static_cast<quint64>(ft.dwHighDateTime) << 32) | ft.dwLowDateTime;
    const qint64 unixMs = static_cast<qint64>(ticks / 10000ULL - 11644473600000ULL);
    nanos = static_cast<int>(ticks % 10000000ULL) * 100; // 100ns units -> ns
    dt = QDateTime::fromMSecsSinceEpoch(unixMs, Qt::UTC).toLocalTime();
}

// JSON number-or-string -> double, missing -> 0.
double getF64(const QJsonObject &v, const QString &key)
{
    const QJsonValue x = v.value(key);
    if (x.isDouble())
        return x.toDouble();
    if (x.isString()) {
        bool ok = false;
        const double d = x.toString().trimmed().toDouble(&ok);
        if (ok)
            return d;
    }
    return 0.0;
}

// JSON number-or-string -> qint64 (fractional/out-of-range doubles rejected,
// mirroring serde_json's Value::as_i64).
std::optional<qint64> getI64(const QJsonValue &v)
{
    if (v.isDouble()) {
        const double d = v.toDouble();
        if (d == std::trunc(d) && d >= -9.0e18 && d <= 9.0e18)
            return static_cast<qint64>(d);
        return std::nullopt;
    }
    if (v.isString()) {
        bool ok = false;
        const qint64 n = v.toString().trimmed().toLongLong(&ok);
        if (ok)
            return n;
    }
    return std::nullopt;
}

// resetTime: RFC3339 first, then the looser shapes DateTimeOffset.TryParse
// accepts (space separator, with/without offset; no offset = local time).
std::optional<QDateTime> parseResetTime(const QString &s)
{
    static const QRegularExpression re(QStringLiteral(
        "^(\\d{4})-(\\d{2})-(\\d{2})[T ](\\d{2}):(\\d{2}):(\\d{2})(\\.\\d+)?\\s*(Z|[+-]\\d{2}:?\\d{2})?$"));
    const QRegularExpressionMatch m = re.match(s.trimmed());
    if (!m.hasMatch())
        return std::nullopt;
    const QDate date(m.captured(1).toInt(), m.captured(2).toInt(), m.captured(3).toInt());
    // Qt is millisecond-based: keep the first three fraction digits
    int msec = 0;
    const QString frac = m.captured(7);
    if (frac.size() > 1)
        msec = (frac.mid(1, 3) + QStringLiteral("000")).left(3).toInt();
    const QTime time(m.captured(4).toInt(), m.captured(5).toInt(), m.captured(6).toInt(), msec);
    if (!date.isValid() || !time.isValid())
        return std::nullopt;
    const QString off = m.captured(8);
    if (off.isEmpty())
        return QDateTime(date, time, Qt::LocalTime);
    if (off == QLatin1String("Z"))
        return QDateTime(date, time, QTimeZone::UTC).toLocalTime();
    const int sign = off.startsWith(QLatin1Char('-')) ? -1 : 1;
    const QString digits = QString(off).remove(QLatin1Char(':')).mid(1);
    const int hours = digits.left(2).toInt();
    const int minutes = digits.mid(2, 2).toInt();
    // Validate like DateTimeOffset.TryParse: offsets beyond +/-14:00 or with
    // minutes > 59 are invalid -> no reset time (Rust returns None).
    if (hours > 14 || minutes > 59)
        return std::nullopt;
    const QTimeZone zone(sign * (hours * 3600 + minutes * 60));
    if (!zone.isValid())
        return std::nullopt;
    return QDateTime(date, time, zone).toLocalTime();
}

QuotaSegment parseSegment(const QJsonValue &v)
{
    // Non-object input behaves like an empty object, as serde Value::get does
    const QJsonObject o = v.isObject() ? v.toObject() : QJsonObject();
    const double used = getF64(o, QStringLiteral("used"));
    double limit = getF64(o, QStringLiteral("limit"));
    if (limit <= 0.0)
        limit = 1.0; // guard against division by zero
    QuotaSegment seg;
    const QJsonValue rt = o.value(QStringLiteral("resetTime"));
    if (rt.isString())
        seg.resetAt = parseResetTime(rt.toString());
    // getF64 can yield inf/NaN from hostile strings ("1e999", "NaN"); keep
    // percent finite so serialization never emits a non-finite double
    const double percent = used / limit * 100.0;
    seg.percent = std::isfinite(percent) ? percent : 0.0;
    return seg;
}

std::optional<qint64> parseCents(const QJsonValue &money)
{
    if (!money.isObject())
        return std::nullopt;
    return getI64(money.toObject().value(QStringLiteral("priceInCents")));
}

ExtraInfo parseExtra(const QJsonValue &wallet)
{
    ExtraInfo info;
    if (!wallet.isObject())
        return info; // not an object -> NotActivated
    const QJsonObject w = wallet.toObject();

    // isEnabled defense: when the booster is disabled, amountLeft is an estimate
    // (monthly limit minus used), NOT the real balance -> must read as NotActivated.
    const QJsonValue enabled = w.value(QStringLiteral("isEnabled"));
    if (enabled.isBool() && !enabled.toBool())
        return info;

    const QJsonValue balance = w.value(QStringLiteral("balance"));
    std::optional<qint64> raw;
    if (balance.isObject())
        raw = getI64(balance.toObject().value(QStringLiteral("amountLeft")));
    if (raw) {
        info.state = ExtraState::Ready;
        // saturating_add: a pathological amountLeft near i64::MAX must not overflow
        const qint64 rounded = *raw > std::numeric_limits<qint64>::max() - 500000
            ? std::numeric_limits<qint64>::max()
            : *raw + 500000;
        info.balanceCents = rounded / 1000000; // 1e-8 yuan -> cents, rounded
    } else {
        info.state = ExtraState::NoData;
    }

    const QJsonValue monthly = w.value(QStringLiteral("monthlyChargeLimitEnabled"));
    if (monthly.isBool() && monthly.toBool()) {
        info.monthlyEnabled = true;
        info.monthlyUsedCents = parseCents(w.value(QStringLiteral("monthlyUsed")));
        info.monthlyLimitCents = parseCents(w.value(QStringLiteral("monthlyChargeLimit")));
    }
    return info;
}

} // namespace

QuotaResult QuotaResult::failed(const QString &kind)
{
    QuotaResult r;
    nowPrecise(r.fetchedAt, r.fetchedAtNanos);
    r.error = kind;
    return r;
}

void QuotaResult::fillMissingFrom(const QuotaResult &last)
{
    if (!fiveHour)
        fiveHour = last.fiveHour;
    if (!week)
        week = last.week;
    if (!extra)
        extra = last.extra;
}

namespace quota {

QuotaResult fetch()
{
    const std::optional<QString> token = credentials::loadToken();
    if (!token)
        return QuotaResult::failed(QStringLiteral("no-token"));

    SyncHttpResult resp = syncGet(
        QUrl(QString::fromLatin1(kUsagesUrl)),
        {{QByteArray("Authorization"), QByteArray("Bearer ") + token->toUtf8()},
         {QByteArray("Accept"), QByteArray("application/json")}});
    if (resp.timedOut)
        return QuotaResult::failed(QStringLiteral("TaskCanceledException"));
    if (!resp.ok || resp.status < 200 || resp.status >= 300)
        return QuotaResult::failed(QStringLiteral("HttpRequestException"));

    QJsonParseError parseError;
    const QJsonDocument doc = QJsonDocument::fromJson(resp.body, &parseError);
    if (parseError.error != QJsonParseError::NoError)
        return QuotaResult::failed(QStringLiteral("JsonException"));
    // Non-object roots parse fine but yield all-null segments, as in Rust
    const QJsonObject root = doc.isObject() ? doc.object() : QJsonObject();

    QuotaResult r;
    nowPrecise(r.fetchedAt, r.fetchedAtNanos);

    // 5-hour segment: root.limits[0].detail
    const QJsonValue limits = root.value(QStringLiteral("limits"));
    if (limits.isArray() && !limits.toArray().isEmpty()) {
        const QJsonValue first = limits.toArray().first();
        const QJsonValue detail =
            first.isObject() ? first.toObject().value(QStringLiteral("detail")) : QJsonValue();
        if (!detail.isUndefined())
            r.fiveHour = parseSegment(detail);
    }
    // Week segment: root.usage
    const QJsonValue usage = root.value(QStringLiteral("usage"));
    if (usage.isObject())
        r.week = parseSegment(usage);
    r.extra = parseExtra(root.value(QStringLiteral("boosterWallet")));
    return r;
}

} // namespace quota
